File: samples/python/src/GraphsPrimitivesEdges.py
from typing import Any

# UI settings
# Expression: gmap
#   Type: dict
#     Layout: Graph
#     Nodes: @dictCustomNodes
#     Save: Object Attributes
#   Type: str
#     Layout: Graph
#     Nodes: @nodeCustomNodes
#     Edges: @nodeCustomEdges
#     Save: Object Type Attributes
class GraphsPrimitivesEdges:
    @staticmethod
    def run_main():
        GraphsPrimitivesEdges().start()

    def start(self):
        # graph with hashmap primitives
        gmap: dict[str, list[list[str]]] = {}
        gmap["1"] = [  # node value
            ["2", "3"],  # children nodes
            ["10", "20"],  # chidlren edges
        ]
        gmap["2"] = [["3", "4"], ["30", "40"]]
        gmap["3"] = []  # no children nodes
        gmap["4"] = []
        gmap["5"] = []

        print("done")


# custom extractor
class Extractor:
    # register the custom attributes
    @staticmethod
    def register_attrs() -> list[list[str, list[str]]]:
        return [
            ["dict", ["dictCustomNodes"]],
            ["str", ["nodeCustomNodes", "nodeCustomEdges"]],
        ]

    @staticmethod
    def extract_dictCustomNodes(
        obj: dict[str, list[list[str]]], root: Any
    ) -> list[str] | list[int] | list[object] | None:
        rootObject: dict[str, list[list[str]]] = root
        nodes: list[str] = []
        for k in rootObject.keys():
            nodes.append(k)
        return nodes

    @staticmethod
    def extract_nodeCustomNodes(
        obj: str, root: dict[str, list[list[str]]]
    ) -> list[list[str]]:
        nodes: list[list[str]] = root.get(obj, [])
        if obj in root and len(root[obj]) >= 1:
            return root[obj][0]

    @staticmethod
    def extract_nodeCustomEdges(
        obj: str, root: dict[str, list[list[str]]]
    ) -> list[str] | list[int] | list[object] | None:
        if obj in root and len(root[obj]) == 2:
            return root[obj][1]


GraphsPrimitivesEdges.run_main()

File: samples/ts/src/GraphsPrimitivesEdges.ts
class GraphsPrimitivesEdges {

    public static runMain() {
        new GraphsPrimitivesEdges().start();
    }

    public start() {
        // It is preferable you use objects since 
        // arrays are used for iterating over node collections.
        // Though you can still use primitive arrays is you wish, see 
        // below an extreme example of a graph using a map with primitives
        // with the help of the Extractor to structure the nodes and edges.

        // UI Settings
        // Expression: gmap

        // Type: Map
        // Layout: Graph
        // Children Nodes: @customNodes

        // Type: Array[][]
        // Layout: Graph
        // Children Nodes: @customNodes
        // Children Edges: @customEdges

        const gmap = new Map<string, [string, string[], string[]]>();
        gmap.set("1", // key
            // value
            ["1", // node
                ["2", "3"], // nodes links
                ["10", "20"] // edges for each link
            ]);
        gmap.set("2", ["2", ["3", "4"], ["30", "40"]]);
        gmap.set("3", ["3", [], []]); // no links
        gmap.set("4", ["4", [], []]);
        gmap.set("5", ["5", [], []]);

        console.log("done");
    }
}

// custom extractor
export class Extractor {
    // register the custom attributes to extract
    public static registerAttrs(): [string, string[]][] {
        return [
            ["Map", ["mapCustomNodes"]],
            ["Array[][]", ["nodeCustomNodes", "nodeCustomEdges", "nodeCustomValue"]],
        ];
    }

    public static extract_mapCustomNodes(
        obj: Map<string, [string, string[], string[]]>,
        root: Map<string, [string, string[], string[]]>
    ): [string, string[], string[]][] {
        const nodes: [string, string[], string[]][] = [];
        for (const value of (root as Map<string, [string, string[], string[]]>).values()) {
            nodes.push(value);
            break;
        }
        return nodes;
    }

    public static extract_nodeCustomNodes(
        obj: [string, string[], string[]],
        root: Map<string, [string, string[], string[]]>
    ): [string, string[], string[]][] {
        const nodes: [string, string[], string[]][] = [];
        if (obj.length >= 2 && obj[1]) {
            for (const id of obj[1]) {
                const nodeObj = root.get(id);
                if (nodeObj) {
                    nodes.push(nodeObj);
                }
            }
        }
        return nodes;
    }
    public static extract_nodeCustomEdges(
        obj: [string, string[], string[]],
        root: Map<string, [string, string[], string[]]>
    ): string[] {
        // edges
        if (obj.length >= 3 && obj[2])
            return obj[2];
        return [];
    }

    public static extract_nodeCustomValue(
        obj: [string, string[], string[]],
        root: Map<string, [string, string[], string[]]>
    ): string[] {
        return [String(obj[0])];
    }
}

GraphsPrimitivesEdges.runMain();